File: Codeforces/126B Password/B_Password-KMP.py
#  Codeforces 126B Password
# https://codeforces.com/problemset/problem/126/B

import os, sys, io, array
sys.stdin = io.StringIO(os.read(0, os.fstat(0).st_size).decode())
input = sys.stdin.readline

s = input().rstrip()

LPS = array.array('i', [0])* len(s)
k = 0

for i in range(1, len(s)):
    while s[i] != s[k] and k:
        k = LPS[k-1]
        
    k += s[i] == s[k]
    
    LPS[i] = k
    
len_boundary = LPS[-1]

if len_boundary == 0:
    print("Just a legend")
    exit(0)
    
for n in LPS[:-1]:
    if n == len_boundary:
        print(s[:len_boundary])
        exit(0)

if LPS[len_boundary-1] == 0:
    print("Just a legend")
    exit(0)

print(s[:LPS[len_boundary-1]])

File: Codeforces/126B Password/B_Password-KMP.cpp
//  Codeforces 126B Password
// https://codeforces.com/problemset/problem/126/B

#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    string s; cin >> s;

    int len_s = s.size();

    vector<int> LPS(len_s);
    int k = 0;
    for(int i = 1; i < len_s; ++i) {
        while(k > 0 and s[i] != s[k]) {
            k = LPS[k - 1];
        }
        k += s[i] == s[k];
        LPS[i] = k;
    }

    int len_boundary = LPS[len_s-1];

    if (len_boundary == 0) {
        cout << "Just a legend\n";
        return 0;
    }

    for( int i = 0; i < len_s-1; ++i) {
        if (LPS[i] == len_boundary) {
            cout << s.substr(0, len_boundary) << endl;
            return 0;
        }
    }



    if (LPS[len_boundary-1] == 0) {
        cout << "Just a legend\n";
        return 0;
    }

    cout << s.substr(0, LPS[len_boundary-1]) << endl;
}
